File: src/main/scala/resonantinduction/atomic/machine/plasma/TilePlasma.java
package resonantinduction.atomic.machine.plasma;

import java.util.ArrayList;

import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.entity.Entity;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.DamageSource;
import net.minecraft.world.IBlockAccess;
import net.minecraftforge.common.util.ForgeDirection;
import net.minecraftforge.common.MinecraftForge;
import org.lwjgl.opengl.GL11;
import resonant.api.event.PlasmaEvent.SpawnPlasmaEvent;
import resonant.content.prefab.java.TileAdvanced;
import resonant.engine.grid.thermal.ThermalGrid;
import resonant.lib.config.Config;
import universalelectricity.core.transform.vector.Vector3;
import universalelectricity.core.transform.vector.VectorWorld;

public class TilePlasma extends TileAdvanced
{
    @Config
    public static int plasmaMaxTemperature = 1000000;
    private float temperature = plasmaMaxTemperature;

    public TilePlasma()
    {
        super(Material.lava);
        textureName_$eq("plasma");
        isOpaqueCube(false);
    }

    @Override
    public int getLightValue(IBlockAccess access)
    {
        return 7;
    }

    @Override
    public boolean isSolid(IBlockAccess access, int side)
    {
        return false;
    }

    @Override
    public ArrayList<ItemStack> getDrops(int metadata, int fortune)
    {
        return new ArrayList<ItemStack>();
    }

    @Override
    public int getRenderBlockPass()
    {
        return 1;
    }

    @Override
    public void collide(Entity entity)
    {
        entity.attackEntityFrom(DamageSource.inFire, 100);
    }

    @Override
    public void update()
    {
        super.update();
        ThermalGrid.addTemperature(new VectorWorld(this), (temperature - ThermalGrid.getTemperature(new VectorWorld(this))) * 0.1f);

        if (ticks() % 20 == 0)
        {
            temperature /= 1.5;

            if (temperature <= plasmaMaxTemperature / 10)
            {
                worldObj.setBlock(xCoord, yCoord, zCoord, Blocks.fire, 0, 3);
                return;
            }

            for (int i = 0; i < 6; i++)
            {
                // Randomize spread direction.
                if (worldObj.rand.nextFloat() > 0.4)
                {
                    continue;
                }

                Vector3 diDian = new Vector3(this);
                diDian.add(ForgeDirection.getOrientation(i));

                TileEntity tileEntity = diDian.getTileEntity(worldObj);

                if (!(tileEntity instanceof TilePlasma))
                {
                    MinecraftForge.EVENT_BUS.post(new SpawnPlasmaEvent(worldObj, diDian.xi(), diDian.yi(), diDian.zi(), (int) temperature));
                }
            }
        }
    }

    public void setTemperature(int newTemperature)
    {
        temperature = newTemperature;
    }

}

File: src/main/scala/resonantinduction/atomic/machine/plasma/TilePlasma.scala
package resonantinduction.atomic.machine.plasma

import java.util.ArrayList

import net.minecraft.block.material.Material
import net.minecraft.entity.Entity
import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.DamageSource
import net.minecraft.world.IBlockAccess
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.common.util.ForgeDirection
import resonant.api.event.PlasmaEvent
import resonant.content.prefab.java.TileAdvanced
import resonant.engine.grid.thermal.ThermalGrid
import resonant.lib.mod.config.Config
import resonant.lib.transform.vector.Vector3

object TilePlasma
{
    @Config var plasmaMaxTemperature: Int = 1000000
}

class TilePlasma extends TileAdvanced(Material.lava)
{
    private var temperature: Double = TilePlasma.plasmaMaxTemperature

    //Constructor
    textureName_$eq("plasma")
    isOpaqueCube(false)

    override def getLightValue(access: IBlockAccess): Int =
    {
        return 7
    }

    override def isSolid(access: IBlockAccess, side: Int): Boolean =
    {
        return false
    }

    override def getDrops(metadata: Int, fortune: Int): ArrayList[ItemStack] =
    {
        return new ArrayList[ItemStack]
    }

    override def getRenderBlockPass: Int =
    {
        return 1
    }

    override def collide(entity: Entity)
    {
        entity.attackEntityFrom(DamageSource.inFire, 100)
    }

    override def update
    {
        super.update
        ThermalGrid.addTemperature(toVectorWorld, ((temperature - ThermalGrid.getTemperature(toVectorWorld)) * 0.1f).asInstanceOf[Float])
        if (ticks % 20 == 0)
        {
            temperature /= 1.5
            if (temperature <= TilePlasma.plasmaMaxTemperature / 10)
            {
                worldObj.setBlock(xCoord, yCoord, zCoord, Blocks.fire, 0, 3)
                return
            }
            for (i <- 0 to 6)
            {
                if (worldObj.rand.nextFloat < 0.4)
                {
                    val diDian: Vector3 = toVector3
                    diDian.add(ForgeDirection.getOrientation(i))
                    val tileEntity: TileEntity = diDian.getTileEntity(worldObj)
                    if (!(tileEntity.isInstanceOf[TilePlasma]))
                    {
                        MinecraftForge.EVENT_BUS.post(new PlasmaEvent.SpawnPlasmaEvent(worldObj, diDian.xi, diDian.yi, diDian.zi, temperature.asInstanceOf[Int]))
                    }
                }
            }
        }
    }

    def setTemperature(newTemperature: Int)
    {
        temperature = newTemperature
    }

}